### DISCUSSION TITLE
Renovate ignores explicitly set onboarding config extends via env/global config

### DISCUSSION THREAD

**Discussion by LukasAuerbeck** [renovatebot/renovate#30202] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

Renovate ignores custom config presets set via the env var / global renovate config.

Assume this renovate config:
```
    module.exports = {
      onboardingConfigFileName: "renovate.json5",
      onboardingRebaseCheckbox: true,
      onboardingNoDeps: true,
      semanticCommits: "enabled",
      detectHostRulesFromEnv: true
    }
```

with this env variable:
```
export RENOVATE_ONBOARDING_CONFIG='{"$$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["local>org/group/renovate-config#v1.26.2"]}'
```

I would assume that the config above creates a new PR with the rendered config
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>org/group/renovate-config#v1.26.2"
  ]
}
```

Instead, the following config gets created:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>org/group/renovate-config"
  ]
}
```

This is because the function [getOnboardingConfig](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/onboarding/branch/config.ts#L12C16-L12C35) ignores whether the onboardingConfig already explicitly sets a renovate-config.

I already opened a PR to fix this, but it was closed since a discussion about the requirements is required first: https://github.com/renovatebot/renovate/pull/30200/

**Comment by rarkins** [comment]
Does it only happen if using env, or also happens if configured in config.js? Does it happen only if you have all the settings you listed in the config.js, or are those irrelevant to this discussion?

**Reply by LukasAuerbeck** [reply]
The onboarding* related settings in the config make it easier to replicate with a demo repo but are not relevant for the issue. It also doesn't matter how the onboarding config gets set, since the env settings and the config.js get combined before.
Configuring the onboarding config via env is just easier since it doesn't require escaping any special characters ($ would get resolved, ...)

**Reply by rarkins** [reply]
Wouldn't you expect Renovate to use the onboardingConfig exactly as defined, even if no extends is included?

**Reply by LukasAuerbeck** [reply]
Yes, but I think an explicitly defined onboardingConfig should have a higher priority if the user configures one.

This is currently not the case. I would only expect renovate to try to find an existing `renovate-config` repo if the `config.js` doesn't explicitly configure an extends, like these configs:

```js
{
  onboardingConfig: {}
}
```

```js
{}
```

For these 2 `config.js` renovate should find the `org/group/renovate-config` repo during the onboarding and create this json:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>org/group/renovate-config"
  ]
}
```

This is the current behaviour and is working as expected.

However, when I configure this `config.js`:

```js
{
  onboardingConfig: {
    extends: [
      "local>org/group/renovate-config#v1.23.2"
    ]
  }
}
```

I would expect the onboarding json to be:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>org/group/renovate-config#v1.23.2"
  ]
}
```

This is what currently is not working, since renovate still iterates through the groups and tries to find a `renovate-config` repo, and if found, overwrites the explicitly configured value.

**Reply by rarkins** [reply]
> I would only expect renovate to try to find an existing `renovate-config` repo if the `config.js` doesn't explicitly configure an extends, like these configs:
> 
> ```js
> {
>   onboardingConfig: {}
> }
> ```

I disagree on this one. If the user has explicitly defined an empty onboardingConfig then we should honor that.

> ```js
> {}
> ```

Here I agree that Renovate should look for an extends to add

> However, when I configure this `config.js`:
> 
> ```js
> {
>   onboardingConfig: {
>     extends: [
>       "local>org/group/renovate-config#v1.23.2"
>     ]
>   }
> }
> ```
> 
> I would expect the onboarding json to be:
> 
> ```json
> {
>   "$schema": "https://docs.renovatebot.com/renovate-schema.json",
>   "extends": [
>     "local>org/group/renovate-config#v1.23.2"
>   ]
> }
> ```
> 
> This is what currently is not working, since renovate still iterates through the groups and tries to find a `renovate-config` repo, and if found, overwrites the explicitly configured value.

I agree on the above

**Reply by LukasAuerbeck** [reply]
ok, I reopened a new PR (https://github.com/renovatebot/renovate/pull/30350) and slightly changed some things:

I think it also makes sense to handle an empty extends like this:

```js
{
  onboardingConfig: {
    extends: []
  }
}
```

to generate an onboarding config like this:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": []
}
```
